packages/loader/container-loader/src/containerContext.ts

```diff
@@ -326,10 +340,16 @@ export class ContainerContext implements IContainerContext {
 
     private async instantiateRuntime(existing: boolean) {
         const runtimeFactory = await this.getRuntimeFactory();
-        this._runtime = await PerformanceEvent.timedExecAsync(
-			this.taggedLogger,
-			{ eventName: "InstantiateRuntime" },
-			async () => runtimeFactory.instantiateRuntime(this, existing));
+        try {
+            this._runtime = await PerformanceEvent.timedExecAsync(
+                this.taggedLogger,
+                { eventName: "InstantiateRuntime" },
+                async () => runtimeFactory.instantiateRuntime(this, existing));
+            this._entryPointPromiseResolver(this._runtime.entryPoint);
```

Comment:
Deferred is an antipattern that is really only necessary here because we don't have firmly defined lifecycle state transitions (instantiating the runtime should be a codified lifecycle state transition).  I'd prefer to try to clarify this state transition - an easy way to do this is to introduce a private EventEmitter that emits on lifecycle state transitions (e.g. here you would `this.lifecycleState.emit("runtimeInstantiated")` or similar.

And then above you can just do something like
```ts
        this.entryPoint = new Promise<FluidObject | undefined>((resolve, reject) => {
            if (this._runtime !== undefined) {
                resolve(this._runtime.entryPoint);
            } else {
                this.lifecycleState.once("runtimeInstantiated", () => { resolve(this._runtime.entryPoint); });
            }
        });
```

Example of this pattern used in a somewhat more complex scenario in Quorum: https://github.com/microsoft/FluidFramework/blob/main/server/routerlicious/packages/protocol-base/src/quorum.ts#L234

Really this is just a symptom of a bigger problem we have in ContainerContext that the instantiation flow is delegated in a very weird way, but this at least avoids the Deferred antipattern without any new public API surface.

Reply:
if you go with this approach it should be a lazy promise. a regular promise will immediately get put into the event queue.

Reply:
Would that be a problem, though? Seems like the promise resolving to the runtime in the same JS turn when it is already there would be fine.

Reply:
TBH I'd go with a non-lazy -- the entrypoint author (container author) can always do their own lazy-fication of anything that is expensive to initialize up-front, but if we lazy-fy on their behalf they don't have a way out of it.

Reply:
`entryPoint` became `getEntryPoint()` in the context, and in the runtime the initialization function is now called on first access to its `entryPoint`, so I think this is somewhat moot. Regular Promise seems fine in the context's `getEntryPoint()` method since a call to it almost certainly means the user wants the entryPoint now and it should be initialized if necessary.